packages/next/src/next-devtools/dev-overlay/components/devtools-panel/devtools-panel.tsx

```diff
@@ -1,12 +1,84 @@
+import type { OverlayDispatch, OverlayState } from '../../shared'
+
+import { Suspense, useRef } from 'react'
+import { Dialog, DialogContent, DialogHeader, DialogBody } from '../dialog'
+import { Overlay } from '../overlay/overlay'
+import { useFocusTrap } from '../errors/dev-tools-indicator/utils'
+import { useDelayedRender } from '../../hooks/use-delayed-render'
+import { ACTION_DEVTOOLS_PANEL_TOGGLE } from '../../shared'
 import { css } from '../../utils/css'
 
-export function DevToolsPanel() {
-  // TODO: Remove style; It is to indicate faster in the browser.
-  return <div data-nextjs-devtools-panel>DevToolsPanel</div>
+export function DevToolsPanel({
+  state,
+  dispatch,
+}: {
+  state: OverlayState
+  dispatch: OverlayDispatch
+}) {
+  const dialogRef = useRef<HTMLDivElement>(null)
+
+  // This hook lets us do an exit animation before unmounting the component
+  const { mounted, rendered } = useDelayedRender(state.isDevToolsPanelOpen)
+
+  useFocusTrap(dialogRef, null, rendered)
+
+  if (!mounted) {
+    // Workaround React quirk that triggers "Switch to client-side rendering" if
+    // we return no Suspense boundary here.
+    return <Suspense />
+  }
+
+  const onClose = () => {
+    dispatch({ type: ACTION_DEVTOOLS_PANEL_TOGGLE })
```

Comment:
Can we only pass the onClose prop to DevtoolPanel? This can be more headless instead of bound with dispatch. We can group the dispatch actions on the top and pass them down as separate props instead of passing the dispatch itself which might lead to abuse in the future